tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/AbstractGradientAdapter.java

```diff
@@ -0,0 +1,93 @@
+/*
+ Copyright 2024 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================
+*/
+package org.tensorflow;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.bytedeco.javacpp.Pointer;
+import org.bytedeco.javacpp.PointerPointer;
+import org.bytedeco.javacpp.PointerScope;
+import org.tensorflow.internal.c_api.TF_GradFuncAdapter;
+import org.tensorflow.internal.c_api.TF_GraphId;
+import org.tensorflow.internal.c_api.TF_Operation;
+import org.tensorflow.internal.c_api.TF_Output;
+import org.tensorflow.internal.c_api.TF_Scope;
+
+/** Helper base class for custom gradient adapters <b>INTERNAL USE ONLY</b> */
```

Comment:
Should we try to put it in a different package to delineate internal vs external API or does it need protected access into `Graph`?

Reply:
It needs access to protected method in `Graph`, yes. That is also why the abstract class is not in the same package as other gradient stuff (`org.tensorflow.op`).

Reply:
Every so often I wish for friend classes in Java. Oh well.